docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,358 @@
+// block.
+message GetLatestBlockIDResponse {
+    // The height of the latest committed block.
+    int64 height = 1;
+    // The ID of the latest committed block.
+    tendermint.types.BlockID block_id = 2;
+}
+
+message GetBlockRequest {
+    // The height of the block to get.
+    int64 height = 1;
+}
+
+message GetBlockResponse {
+    // Block data for the requested height.
+    tendermint.types.Block block = 1;
+}
+
+message GetBlockResultsRequest {
+    // The height of the block results to get.
+    int64 height = 1;
+}
+
+message GetBlockResultsResponse {
+    // All events produced by the ABCI BeginBlock call for the block.
+    repeated tendermint.abci.Event begin_block_events = 1;
+
+    // All transaction results produced by block execution.
+    repeated tendermint.abci.ExecTxResult tx_results = 2;
+
+    // Validator updates during block execution.
+    repeated tendermint.abci.ValidatorUpdate validator_updates = 3;
+
+    // Consensus parameter updates during block execution.
+    tendermint.types.ConsensusParams consensus_param_updates = 4;
+
+    // All events produced by the ABCI EndBlock call.
+    // NB: This should be called finalize_block_events when ABCI 2.0 lands.
+    repeated tendermint.abci.Event end_block_events = 5;
+}
+```
+
+#### Data Companion Service
+
+```protobuf
+syntax = "proto3";
+
+package tendermint.data_companion_service.v1;
+
+// DataCompanionService provides privileged access to specialized pruning
+// functionality on the Tendermint node to help optimize node storage.
+service DataCompanionService {
+    // SetRetainHeight notifies the node of the minimum height whose data must
+    // be retained by the node. This data includes block data and block
+    // execution results.
+    //
+    // Setting a retain height lower than a previous setting will result in an
+    // error.
+    //
+    // The lower of this retain height and that set by the application in its
+    // Commit response will be used by the node to determine which heights' data
+    // can be pruned.
+    rpc SetRetainHeight(SetRetainHeightRequest) returns (SetRetainHeightResponse) {}
+
+    // GetRetainHeight returns the retain height set by the companion and that
+    // set by the application. This can give the companion an indication as to
+    // which heights' data are currently available.
+    rpc GetRetainHeight(GetRetainHeightRequest) returns (GetRetainHeightResponse) {}
+}
+
+message SetRetainHeightRequest {
+    int64 height = 1;
+}
+
+message SetRetainHeightResponse {}
+
+message GetRetainHeightRequest {}
+
+message GetRetainHeightResponse {
+    // The retain height as set by the data companion.
+    int64 data_companion_retain_height = 1;
+    // The retain height as set by the ABCI application.
+    int64 app_retain_height = 2;
+}
+```
+
+With this API design, it is technically possible for an integrator to attach
+multiple data companions to the node, but only one of their retain heights will
+be applied.
```

Comment:
So, semantically (as far as the interface is concerned) it is still "at most one"